config/src/lib.rs

```diff
@@ -56,106 +80,201 @@ pub struct ConfigOption {
-pub fn get_matcher<'a>(
-    config_description: &dyn ConfigDescription,
-    config_file_env: &str,
-    args: Vec<OsString>,
-) -> Result<ArgMatches<'a>, Error> {
-    let configuration = config_description.get_config();
-    let options = &configuration.options;
-
-    let mut matches = App::new(configuration.name)
-        .version(configuration.version)
-        .about(configuration.about);
-
-    for option in options.iter() {
-        matches = matches.arg(
-            Arg::with_name(option.name)
+struct ConfigBuilder {}
+
+impl ConfigBuilder {
+    /// The entry point into ConfigBuilder, this method will be called by any binary (or library)
+    /// using this lib for command line parsing.
+    /// It is a typed method, with the type having to implement the Configurable trait, so the type
+    /// has to offer two methods: one to retrieve the definition of command line parameters, one to
+    /// pass back the parsed values of those parameters.
+    ///
+    /// * `commandline` The command line parameters to parse the configuration from (first element will be
+    /// ignored, as this is the binary name
+    /// * `config_file_env` Name of the environment variable to read an extra config file from
+    pub fn build<T: Configurable>(
+        commandline: Vec<OsString>,
+        config_file_env: &str,
+    ) -> Result<T, Error> {
+        // Parse commandline according to config definition
+        let description = T::get_config_description();
+
+        // Use the command line parameters defined in the description to build a
+        // clap matcher object that can be used to parse the acual parameters
+        let matcher = ConfigBuilder::create_matcher(&description);
+
+        // Overwrite command line arguments with final arguments to parse
+        // if a config file was specified, all options from that file will be
+        // prepended to the command line arguments
+        let commandline =
+            ConfigBuilder::maybe_combine_arguments(matcher.clone(), commandline, config_file_env);
+
+        // Parse command line
+        let matcher = matcher.get_matches_from(commandline.expect("Error parsing commandline!"));
+
+        // Convert results from command line parsing into a HashMap<ConfigOption, Vec<String>>
+        // this is then passed to the actual implementation of the configuration for processing
+        let mut result: HashMap<ConfigOption, Option<Vec<String>>> = HashMap::new();
+        for config_option in description.options.clone() {
+            println!(
+                "{}: is_present(): {} - occurrences_of(): {}",
+                config_option.name,
+                matcher.is_present(config_option.name),
+                matcher.occurrences_of(config_option.name)
+            );
+            if matcher.occurrences_of(config_option.name) != 0 {
+                let found_values = matcher
+                    .values_of(config_option.name)
+                    .expect("error")
```

Comment:
More descriptive error message?
I know it shouldn't happen because you already checked if it's available but maybe change to at least "Should never happen" or something like that.

Reply:
Added proper error message.